Build 21-0845 front end

tlei123 opened the pull request:
## Summary

- Build front-end chapters & pages for Authorize VA to Release PI (21-0845)
- Team: VA Product Forms
- Flipper `form2110210` enabled only in Staging; form not launched yet.

## Related issue(s)

- _Link to ticket created in va.gov-team-forms repo_
department-of-veterans-affairs/va.gov-team-forms#297
- _Link to previous change of the code/bug (if applicable)_
[previous relevant vets-website PR](https://github.com/department-of-veterans-affairs/vets-website/pull/24374)

## Testing done

- Local browser testing; automated tests coming soon.

## What areas of the site does it impact?

This WIP form only.

tlei123 committed:
Scaffold new 21-0845 form-app

tlei123 committed:
Merge branch 'main' into va-forms/21-0845/297-build-front-end

tlei123 committed:
Finish 21-0845 Intro & Confirmation pages

tlei123 committed:
Merge branch 'main' into va-forms/21-0845/297-build-front-end

tlei123 committed:
Add 21-0845 vet-personal-info page

tlei123 committed:
Merge branch 'main' into va-forms/21-0845/297-build-front-end

tlei123 committed:
Merge branch 'main' into va-forms/21-0845/297-build-front-end

tlei123 committed:
Merge branch 'main' into va-forms/21-0845/297-build-front-end

tlei123 committed:
Merge branch 'main' into va-forms/21-0845/297-build-front-end

tlei123 committed:
Finish 21-0845 Intro page

tlei123 committed:
Finish 21-0845 Confirmation page

tlei123 committed:
Merge branch 'main' into va-forms/21-0845/297-build-front-end

tlei123 committed:
Finish 21-0845 authorizer-type page

tlei123 committed:
Merge branch 'main' into va-forms/21-0845/297-build-front-end

tlei123 committed:
Move 21-0845 constants file

tlei123 committed:
Dynamicize 21-0845 veteran-personal-information page step-title

tlei123 committed:
Add more pages to 21-0845

tlei123 committed:
Merge branch 'main' into va-forms/21-0845/297-build-front-end

tlei123 committed:
browserslist/caniuse db-update

tlei123 committed:
Add more pages to 21-0845

tlei123 committed:
Add info-scope page to 21-0845

tlei123 committed:
Rename 21-0845 test-data file

tlei123 committed:
Add 21-0845 limited-info page

tlei123 committed:
Add release-duration page to 21-0845

tlei123 committed:
Fix 21-0845 limited-info page styling

tlei123 committed:
Merge branch 'main' into va-forms/21-0845/297-build-front-end

tlei123 committed:
Add 21-0845 release-end-date page

tlei123 committed:
Add 21-0845 security-question & -answer pages

tlei123 committed:
Refactor 21-0845 page-files constants usage

tlei123 committed:
Add 21-0845 authorizer-info pages

tlei123 committed:
Merge branch 'main' into va-forms/21-0845/297-build-front-end

# Conflicts [resolved]:
#	yarn.lock

tlei123 committed:
Merge branch 'main' into va-forms/21-0845/297-build-front-end

tlei123 committed:
Update 21-0845 content further

tlei123 committed:
Merge branch 'main' into va-forms/21-0845/297-build-front-end

tlei123 committed:
Merge branch 'main' into va-forms/21-0845/297-build-front-end

tlei123 committed:
Merge branch 'main' into va-forms/21-0845/297-build-front-end

tlei123 committed:
Fix RadioWidget false-positive console-error for invalid prop type

tlei123 commented on src/applications/edu-benefits/1990/config/form.js:
```diff
@@ -17,6 +17,8 @@ import preSubmitInfo from 'platform/forms/preSubmitInfo';
 import { VA_FORM_IDS } from 'platform/forms/constants';
 import currentOrPastMonthYearUI from 'platform/forms-system/src/js/definitions/currentOrPastMonthYear';
 import yearUI from 'platform/forms-system/src/js/definitions/year';
+import { validateBooleanGroup } from 'platform/forms-system/src/js/validation';
```
Prob'ly some linter auto-fix that snuck in from a merge from main.

tlei123 commented on src/applications/edu-benefits/1990/config/form.js:
```diff
@@ -17,6 +17,8 @@ import preSubmitInfo from 'platform/forms/preSubmitInfo';
 import currentOrPastMonthYearUI from 'platform/forms-system/src/js/definitions/currentOrPastMonthYear';
 import yearUI from 'platform/forms-system/src/js/definitions/year';
+import { validateBooleanGroup } from 'platform/forms-system/src/js/validation';
+import dateUI from 'platform/forms-system/src/js/definitions/date';
```
Prob'ly some linter auto-fix that snuck in from a merge from main.

tlei123 commented on src/platform/forms-system/src/js/widgets/RadioWidget.jsx:
```diff
@@ -99,7 +99,7 @@ RadioWidget.propTypes = {
     selectedProps: PropTypes.shape({}),
     enableAnalytics: PropTypes.bool,
-    title: PropTypes.string,
+    title: PropTypes.string || PropTypes.element,
```
Gets rid of the false-positive console-error for invalid prop-type when providing JSX for the title.

tlei123 commented on src/platform/forms/save-in-progress/SaveInProgressIntro.jsx:
```diff
@@ -61,6 +61,7 @@ class SaveInProgressIntro extends React.Component {
     let alert;
     let includesFormControls = false;
     const {
+      alertTitle,
```
New optional prop for custom title [see default value near bottom of file].

tlei123 commented on src/applications/simple-forms/shared/components/ConfirmationPageView.jsx:
```diff
@@ -7,6 +7,8 @@ import { waitForRenderThenFocus } from 'platform/utilities/ui';
 
 export const ConfirmationPageView = ({
+  formType = 'application',
+  submitterHeader = 'Applicant',
```
New optional props for type of form and applicant title.

rhasselle-oddball commented on src/applications/simple-forms/21-0845/components/GroupCheckboxWidget.jsx:
```diff
@@ -0,0 +1,76 @@
+import React, { useState } from 'react';
```
this is the same as the one from 21-10210 right? what do you think of moving this to our simple-forms/shared/components and using in both places?

tlei123 replied:
Moved.  I'll open a follow-up PR for 21-10210's move soon.

rhasselle-oddball commented on src/applications/simple-forms/21-0845/config/form.js:
```diff
@@ -1,14 +1,40 @@
+import getHelp from '../../shared/components/GetFormHelp';
+import { AUTHORIZER_TYPES } from '../definitions/constants';
+// pages
+import authTypePg from '../pages/authType';
```
abbreviations is a good idea to save some characters if they are standard/widely recognized abbreviations, but
some abbreviations is not obvious what they could mean, so spelling it out is helpful, for example "rel", "sec".
non-standard abbreviations also feel a little strange to abbreviate "pers" or "Pg", when normally you would use "person" or "page" here.
"auth" can sometimes stand for authorization or authentication.
so I think spelling out things completely has some benefit, especially when you revisit the code in the future and don't remember the context.

tlei123 replied:
Refactored page files/vars.

rhasselle-oddball commented on src/applications/simple-forms/21-0845/definitions/form-address.js:
```diff
@@ -0,0 +1,338 @@
+// copy of src/platform/forms/definitions/address.js,
```
this is same as 21-10210 right? could also move this to simple-forms/shared/definitions.

tlei123 replied:
Refactored & moved to simple-forms/shared/definitions.

tlei123 committed:
Merge branch 'main' into va-forms/21-0845/297-build-front-end

rhasselle-oddball commented on src/applications/simple-forms/21-0845/sass/0845-auth-disclose.scss:
```diff
@@ -5,3 +5,29 @@
 @import "~@department-of-veterans-affairs/formation/sass/modules/m-omb-info";
 @import "../../../../platform/forms/sass/m-form-confirmation";
+
+div:has(.schemaform-sip-alert) {
```
should this be applied to all save in progress? (could add in the html), or only our form? (css here)

tlei123 replied:
No, it's "custom" spacing just for this form.

rhasselle-oddball commented on src/applications/simple-forms/21-0845/definitions/form-definitions.js:
```diff
@@ -0,0 +1,31 @@
+      first: {
+        type: 'string',
+        minLength: 1,
+        maxLength: 12, // PDF-specific override
```
Let's ask Jeana/Andrew about this approach.

tlei123 replied:
Jeana just said we should set PDF-specific maxLengths going fwd.

rhasselle-oddball commented on src/applications/simple-forms/21-0845/pages/authAddr.js:
```diff
@@ -0,0 +1,16 @@
+    type: 'object',
+    required: ['authorizerAddress'],
+    properties: {
+      authorizerAddress: schema({ definitions }, true),
```
hm, do we want to pass in all the definitions here? maybe its a non-issue to have more definitions than you need

tlei123 replied:
Yes, I think the reason why our address def wants all the definitions from definitions.json passed in is, it can set a default to select "address" from the definitions, and let devs override it if they need "usAddress" or "profileAddress.

rhasselle-oddball commented on src/applications/simple-forms/21-0845/pages/authType.js:
```diff
@@ -0,0 +1,58 @@
+    authorizerType: {
+      'ui:title': (
+        <span
+          className="vads-u-font-family--serif vads-u-font-size--h3 vads-u-font-weight--bold"
```
i forget what we decided to do here, h2 styled as an h3?  if so couldn't we just use the h2 tag?  I remember Jeana had notes on how to do this.

tlei123 replied:
No, this can't be an `<h2/>` because it would screw up the Review-page's A11y content-hierarchy [the accordions are `<h3/>`s.  The _parent_ `<legend/>` for my custom-span needs to stay in place there.

rhasselle-oddball commented on src/applications/simple-forms/21-0845/pages/infoScope.js:
```diff
@@ -0,0 +1,57 @@
+
+          return {
+            title: (
+              <span className="vads-u-font-family--serif vads-u-font-size--h4 vads-u-font-weight--bold">
```
any time we're working with a title, I think we need an h tag rather than a span, and then just style it how we want.

tlei123 replied:
For the same reason as above for authType page, this also needs to be a `<span/>` to avoid screwing up the Review-page's A11y content-hierarchy.

rhasselle-oddball reviewed (commented):
lots of stuff going on in this form! all the core stuff looks great. left a few refactor type comments.

tlei123 committed:
Rename/refactor 21-0845 page filenames/variables

tlei123 committed:
Move 21-0845 GroupCheckboxWidget to shared form-app

tlei123 committed:
Fix 21-0845 Review-page styling

gmrabian reviewed (commented):
Looking good! 
I tested it out locally and didn't notice anything major. Two comments
1. The representative array seemed a bit buggy on the review page. Each rep has an Edit button but it doesn't actually allow you to edit. And the add button only lets you edit the new rep once you've selected the top level Edit option.
<img width="593" alt="Screenshot 2023-06-22 at 4 04 13 PM" src="https://github.com/department-of-veterans-affairs/vets-website/assets/57802560/47e7e7db-ee6a-4267-b6b3-c54df3a4472f">
2. No signature box on review page. This isn't blocking since you can always add it later, just mentioning in case you wanted to include it

tlei123 committed:
Make 21-0845 PDF-specific address-definition shareable

tlei123 committed:
Rename shared pdf-address definition

tlei123 committed:
Make 21-0845 PDF-specific fullNameNoSuffix definition shareable

tlei123 committed:
Delete 21-0845 pdf-specific form-definitions

tlei123 commented:
@rhasselle-oddball , please re-review & approve.

I've made most of your request changes, and left comments for the 2 requested changes that can't be made.

**Requested changes made**:
- Renamed page-files/-variables to eliminate abbreviation-confusion
- Moved GroupCheckboxWidget to .../shared/components
- Moved PDF-specific definitions to .../shared/definitions

**Requested changes not made** [see my Comments in our convos above]:
- custom-span for authType page [now renamed authorizerType.js (authorizerTypePg in form.js)]
- custom-span on infoScope page

tlei123 committed:
Merge branch 'main' into va-forms/21-0845/297-build-front-end

tlei123 committed:
Hide 21-0845 Review-page org-reps premature edit-buttons

tlei123 commented:
> 1. The representative array seemed a bit buggy on the review page.

@gmrabian , please re-review & approve.  I just hid the "premature" edit/add buttons\* which were breaking the Review-page update logic, leaving only the top (page-level) Edit button which works.  Once the panel's in Edit-mode, everything works as expected.

\*We can chat about fixing this properly in the existing forms-library or component if you like.  I'm personally wary of regressions though.

gmrabian reviewed (approved):
Changes related to my comments look good. thanks, Tze!
